compiler/ast/parser.nim

```diff
@@ -2063,38 +2063,41 @@ proc parseTypeClass(p: var Parser): PNode =
 
 proc parseTypeDef(p: var Parser): PNode =
   #|
-  #| typeDef = identWithPragmaDot genericParamList? '=' optInd typeDefAux
-  #|             indAndComment? / identVisDot genericParamList? pragma '=' optInd typeDefAux
+  #| typeDef = identWithPragmaDot '=' optInd typeDefAux
+  #|             indAndComment? / identVisDot genericParamList? pragma? '=' optInd typeDefAux
   #|             indAndComment?
   result = newNodeP(nkTypeDef, p)
-  let identifier = p.identVis(allowDot=true)
-  var identPragma = identifier
-  var foundPragmas = false
-
-  if p.tok.tokType == tkCurlyDotLe:
-    let pragma = optPragmas(p)
-    identPragma = newTreeI(nkPragmaExpr, p.lineInfo, [identifier, pragma])
-    foundPragmas = true
-
-  let genericParam = if p.validInd and p.tok.tokType == tkBracketLe:
-                   if foundPragmas:
-                     # Deprecated since v0.20.0
-                     p.localError ParserReport(kind: rparPragmaBeforeGenericParameters)
-                   parseGenericParamList(p)
-                 else:
-                   p.emptyNode
-
-  if not foundPragmas:
-    let pragma = optPragmas(p)
-    if pragma.kind != nkEmpty:
-      identPragma = newTreeI(nkPragmaExpr, p.lineInfo, [identifier, pragma])
-  elif p.tok.tokType == tkCurlyDotLe:
-    p.localError ParserReport(kind: rparPragmaAlreadyPresent)
-
-  result.add identPragma
-  result.add genericParam
+  let 
+    identifier = p.identVis(allowDot=true)
+    (genericParams, pragmas) =
+      case p.tok.tokType
+      of tkBracketLe:
+        (
+          (if p.validInd: parseGenericParamList(p) else: p.emptyNode),
+          optPragmas(p)
+        )
+        # xxx: should the else/empty generic branch invalid indentation error?
+      of tkCurlyDotLe:
+        let res = (p.emptyNode, optPragmas(p))
+        
+        if p.validInd and p.tok.tokType == tkBracketLe:
+          p.localError ParserReport(kind: rparPragmaBeforeGenericParameters)
+        
+        res
+      of tkEquals:
+        (p.emptyNode, p.emptyNode)
+        # parsing for everything after the equals is shared, we do that below
+      else:
+        (p.emptyNode, p.emptyNode) # xxx: error?
```

Comment:
The `else` case is handled by sem, `type abc` on it's own is valid in the parser, so I'd merge the `tkEquals` and the `else` branch. If we don't want/need a specialized error message for "pragma before generic param" then we can also remove the tkCurlyDotLe branch and have it handled with the generic "unexpected X" automatically I think.

Reply:
I prefer to start with a focused grammar, then loosen restrictions over time (reasoning for all this is in the 'out loud thinking' below). So having the grammar restricted in the parser seems reasonable and I'd like to introduce more of that.

I likely don't understand the art of grammar construction well enough, so I'm wondering if it's reasonable to say that `type abc` shouldn't be valid here hence the `else` being separate?

As for the special case error message for pragma and position ordering, that's friendlier to humans.

----
(Out loud thinking)

I'm still getting a feel for what should be where, in particular a set of principles in place.

If we know there is an irregularity in the grammar we should reject it early, IMO. We get far more helpful error messages that way in that we've got better context.

There are a bunch of competing ideas in my head, but they boil down to two main "sides":

1. the focused grammar of Nimskull the language itself and building to that, with crisp and clear error messages, quick detection of issues and syntax limiting is more useful, etc
2. a superset grammar, that can represent all legal terms the former does _and_, is more forgiving for the purposes of building DSLs and other things macros should let you do. All within reason, not full reader macro control like lisp/racket.

The focused route (1) has lots of upsides, not the least of which is simplicity in how direct it is -- consider how much people struggled with `|`, which is alternation in regular grammars.

1 and 2 are somewhat mutually exclusive, we can start with one and move to 2, but the other way is much harder. We can also have 1 be the norm and in special sections drop to 2. In the end we need to pick a horse.

I started to sketch out grammar structures for proc and type definitions in my notepad to see what the symmetries are, mostly ignoring current syntax. I'm starting to get some insights, but it's very nascent and even if it was ready as an idea, it's an entirely different matter implementing it.